### PR TITLE
[pinot] Adding query option support for DynamicTable

### PR DESCRIPTION
**Pull request by xiangfu0**
## Description
Adopted the idea introduced by (https://github.com/trinodb/trino/pull/17001#issuecomment-1594391322),
Support QueryOptions in Pinot DynamicTable.

label: pinot

## Additional context and related issues
Before:
image
After:
image

Server logs: 
```
2023-09-24T16:01:57.264-0700	INFO	20230924_230157_00003_92mfq.0.0.0-0-112	io.trino.plugin.pinot.client.PinotClient	Query 'SET skipUpsert = 'true';
SELECT "event_id", count(*) FROM upsertMeetupRsvp GROUP BY "event_id" LIMIT 10' on broker host 'http://192.168.1.248:8000/query/sql'
```

## Release notes

(X) Release notes are required, with the following suggested text:

```markdown
# Pinot
* Add support for [query options](https://docs.pinot.apache.org/users/user-guide-query/query-options) in Dynamic tables
```

### CODE REVIEW

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java [hunk 1]
```diff
@@ -448,4 +448,25 @@ public void testQuotesInColumnName()
...
+                SELECT FlightNum
+                FROM %s
+                LIMIT 50;
+                SET skipUpsert=true;""".formatted(tableNameWithSuffix);
+        DynamicTable dynamicTable = buildFromPql(pinotMetadata, new SchemaTableName("default", query), mockClusterInfoFetcher, TESTING_TYPE_CONVERTER);
+        String expectedPql = """
+                SELECT "FlightNum" \
+                FROM %s \
+                LIMIT 50;
+                SET skipUpsert = true;
```
Comment (ebyhr) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java:467]: Does Pinot use this option in the above SELECT statement even in this order (SELECT → SET)?

Comment (xiangfu0) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java:467]: Order doesn't matter in a multi statement query

Comment (xiangfu0) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java]: updated the test to move SET statement front

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java [hunk 2]
```diff
@@ -151,7 +151,7 @@ public List<String> listSchemaNames(ConnectorSession session)
     @Override
     public PinotTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
     {
-        if (tableName.getTableName().trim().startsWith("select ")) {
+        if (tableName.getTableName().trim().contains("select ")) {
```
Comment (elonazoulay) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java:154]: This would allow table names like "fooselect ", etc. which would be a syntax error. Why was this changed?

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java [hunk 3]
```diff
@@ -448,4 +448,28 @@ public void testQuotesInColumnName()
...
+                FROM %s
+                LIMIT 50;
+                """.formatted(tableNameWithSuffix);
+        DynamicTable dynamicTable = buildFromPql(pinotMetadata, new SchemaTableName("default", query), mockClusterInfoFetcher, TESTING_TYPE_CONVERTER);
+        String expectedPql = """
+                SELECT "FlightNum" \
+                FROM %s \
+                LIMIT 50;
+                SET useMultistageEngine = true;
+                SET skipUpsert = true;
```
Comment (elonazoulay) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java:470]: Does moving the `set` statements only enable those options after the query is executed?

**Review by elonazoulay — COMMENTED** [review]
lgtm % some questions. Thanks for doing this! Users have been asking for this feature:)

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java (hunk 3 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java:470]: In one multi-line query statement, SET statements will be executed first on matter the ordering.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java (hunk 2 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java:154]: Yeah, my feeling is that as long as it's a sql statement, then it's still going to the same code block for execution. We can add a query parser to fail fast, but not a must.

**Review by elonazoulay — APPROVED** [review]

**Comment by xiangfu0** [comment]
> lgtm % some questions. Thanks for doing this! Users have been asking for this feature:)

Thanks!

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java [hunk 4]
```diff
@@ -76,6 +81,7 @@ public DynamicTable(
         this.orderBy = ImmutableList.copyOf(requireNonNull(orderBy, "orderBy is null"));
         this.limit = requireNonNull(limit, "limit is null");
         this.offset = requireNonNull(offset, "offset is null");
+        this.queryOptions = ImmutableMap.copyOf(queryOptions);
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java:84]: Add `requireNonNull`.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java (hunk 3 above)
Comment (ebyhr) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java:470]: I would recommend moving SET to first in DynamicTablePqlExtractor or leave a comment in this test. 2/2 reviewers asked the same question. It means the current logic is confusing.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java [hunk 5]
```diff
@@ -83,6 +84,16 @@ public static String extractPql(DynamicTable table, TupleDomain<ColumnHandle> tu
             }
             builder.append(table.getLimit().getAsLong());
         }
+        if (!table.getQueryOptions().isEmpty()) {
+            builder.append(";\n");
+            for (Map.Entry<String, String> queryOption : table.getQueryOptions().entrySet()) {
+                builder.append("SET ")
+                        .append(queryOption.getKey())
+                        .append(" = ")
+                        .append(queryOption.getValue())
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java:93]: The official Pinot documentation contains single-quotes for string literal. Is there no need to quote for such cases? 
```sql
SET key1 = 'value1';
SET key2 = 123;
SELECT * FROM myTable
```

Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java]: Yes, updated with single quotes.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java (hunk 4 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java]: added.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java (hunk 3 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java]: done.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java [hunk 6]
```diff
@@ -33,6 +34,14 @@ private DynamicTablePqlExtractor()
     public static String extractPql(DynamicTable table, TupleDomain<ColumnHandle> tupleDomain)
     {
         StringBuilder builder = new StringBuilder();
+        Map<String, String> queryOptions = table.getQueryOptions();
+        queryOptions.keySet().stream().sorted().forEach(
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java:38]: Why do we need to sort? Why not use `queryOptions.forEach((key, value) -> ...` ?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java [hunk 7]
```diff
@@ -33,6 +34,14 @@ private DynamicTablePqlExtractor()
...
+                key -> builder
+                        .append("SET ")
+                        .append(key)
+                        .append(" = ")
+                        .append(String.format("'%s'", queryOptions.get(key)))
```
Comment (ebyhr) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java:43]: static import `format` method.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java [hunk 8]
```diff
@@ -448,4 +448,27 @@ public void testQuotesInColumnName()
...
+                SET skipUpsert='true';
+                SET useMultistageEngine='true';
+                SELECT FlightNum
+                FROM %s
+                LIMIT 50;
+                """.formatted(tableNameWithSuffix);
+        DynamicTable dynamicTable = buildFromPql(pinotMetadata, new SchemaTableName("default", query), mockClusterInfoFetcher, TESTING_TYPE_CONVERTER);
+        String expectedPql = """
+                SET skipUpsert = 'true';
+                SET useMultistageEngine = 'true';
```
Comment (ebyhr) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java:467]: Does Pinot accept sting literals for numerics options? e.g. `numReplicaGroupsToQuery`

Can we add an integration test in BasePinotConnectorSmokeTest?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java (hunk 6 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java:38]: This is to make the generated set key-value pair statements ordering fixed.
The ordering itself doesn't matter.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java (hunk 8 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestDynamicTable.java:467]: added test for query option.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java (hunk 7 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java]: done.
